emjapi-samples-jms-p2p/src/main/java/com/sap/xbem/sample/sapcp/jms/p2p/config/TokenRequest.java

```diff
@@ -0,0 +1,326 @@
+        }
+    }
+
+    static class OAuthHandler {
+
+        private final OAuthHandler.AuthSettings authSettings;
+
+        OAuthHandler(OAuthHandler.AuthSettings authSettings) {
+            this.authSettings = authSettings;
+        }
+
+        String doTokenRequest() throws IOException {
+            String oauthFlow = this.authSettings.getoAuthGrantType();
+            if ("password".equalsIgnoreCase(oauthFlow)) {
+                throw new UnsupportedOperationException("Found unknown oauth flow value: " + oauthFlow);
+            } else if ("client_credentials".equalsIgnoreCase(oauthFlow)) {
+                String url = this.authSettings.getoAuthEndpoint();
+                String clientId = this.authSettings.getoAuthClientId();
+                String clientSecret = this.authSettings.getoAuthClientSecret();
+                return this.doTokenRequestClientFlow(url, clientId, clientSecret);
+            } else {
+                throw new IllegalStateException("Found unknown oauth flow value: " + oauthFlow);
+            }
+        }
+
+        private String doTokenRequestClientFlow(String url, String clientId, String clientSecret) throws IOException {
+            String body = String
+                    .format("client_id=%s&client_secret=%s&grant_type=client_credentials&response_type=token", clientId, clientSecret);
+            Map<String, String> headers = new HashMap();
+            headers.put("Content-Type", "application/x-www-form-urlencoded");
+            headers.put("Accept", "application/json");
+            headers.put("User-Agent", "emjapi");
+            String cid = clientId + ":" + clientSecret;
+            String b64 = Base64.getEncoder().encodeToString(cid.getBytes(StandardCharsets.ISO_8859_1));
+            headers.put("Authorization", "Basic " + b64);
+            String postUrl;
+            if (url.endsWith("/oauth/token")) {
+                postUrl = url + "?grant_type=client_credentials";
+            } else {
+                postUrl = url + "/oauth/token" + "?grant_type=client_credentials";
+            }
+
+            String response = this.post(postUrl, body, headers);
+            return this.extractToken(response);
+        }
+
+        String extractToken(String response) throws IOException {
+            int index = response.indexOf("\"access_token\"");
+            int tokenIndex = response.indexOf(34, index + 14) + 1;
+            int lastTokenIndex = response.indexOf(34, tokenIndex);
+            if (index > 0 && lastTokenIndex > tokenIndex) {
+                return response.substring(tokenIndex, lastTokenIndex);
+            } else {
+                throw new IOException("Unable to extract access_token from response: " + response);
+            }
+        }
+
+        private String post(String url, String content, Map<String, String> additionalHeaders) throws IOException {
+            HttpURLConnection con = this.openConnection(url);
```

Comment:
We could use the web client of spring to do the oauth request instead of the raw http url connection. Makes code less.